js/plugins/mcp/src/client/manager.ts

```diff
@@ -0,0 +1,345 @@
+      logger.info(
+        `[MCP Manager] Restarting connection to MCP server '${serverName}' in manager '${this.name}'`
+      );
+      try {
+        await client.restart();
+      } catch (e) {
+        client.disable();
+        this.setError(serverName, {
+          message: `[MCP Manager] Error restarting to server ${serverName}`,
+          detail: `Details: ${e}`,
+        });
+      }
+    }
+  }
+
+  /**
+   * Updates the connections based on a provided map of server configurations.
+   * - Connects any new servers defined in `mcpServers`.
+   * - Disconnects any servers currently connected but not present in `mcpServers`.
+   * - Reconnects existing servers if their configuration appears to have changed (implicitly handled by `connectServer`).
+   * Sets the client's ready state once all connection attempts are complete.
+   * @param mcpServers A record mapping server names to their configurations.
+   */
+  updateServers(mcpServers: Record<string, McpServerConfig>) {
+    this._ready = false;
+    const newServerNames = new Set(Object.keys(mcpServers));
+    const currentServerNames = new Set(Object.keys(this._clients));
+
+    const promises: Promise<void>[] = [];
+    for (const serverName in mcpServers) {
+      promises.push(this.connect(serverName, mcpServers[serverName]));
+    }
+
+    // Disconnect servers that are no longer in the config
+    for (const serverName of currentServerNames) {
+      if (!newServerNames.has(serverName)) {
+        this.disconnect(serverName);
+      }
+    }
+
+    Promise.all(promises)
+      .then(() => {
+        this._ready = true;
+        while (this._readyListeners.length) {
+          this._readyListeners.pop()?.resolve();
+        }
+      })
+      .catch((err) => {
+        while (this._readyListeners.length) {
+          this._readyListeners.pop()?.reject(err);
+        }
+      });
+  }
+
+  /**
+   * Retrieves all tools from all connected and enabled MCP clients managed by this instance.
+   * This method waits for the manager to be ready (all initial connection attempts made)
+   * before fetching tools.
+   *
+   * It iterates through each managed `GenkitMcpClient`, and if the client is not disabled,
+   * it calls the client's `getTools` method to fetch its available tools.
+   * These are then aggregated into a single array.
+   *
+   * This is useful for dynamically providing a list of all available MCP tools to Genkit,
+   * for example, when setting up a Genkit plugin.
+   *
+   * ```ts
+   * const mcpManager = createMcpManager({ ... });
+   * // In your Genkit configuration:
+   * // const allMcpTools = await mcpManager.getActiveTools(ai);
+   * // Then, these tools can be used or registered with Genkit.
+   * ```
+   *
+   * @param ai The Genkit instance, used by individual clients to define dynamic tools.
+   * @returns A Promise that resolves to an array of `ToolAction` from all active MCP clients.
+   */
+  async getActiveTools(ai: Genkit): Promise<ToolAction[]> {
```

Comment:
Add the `{resourceTools?: boolean}` here as well.

Reply:
Done.